opendevin/controller/agent_controller.py

```diff
@@ -87,10 +91,17 @@ async def close(self):
     def update_state_before_step(self):
         self.state.iteration += 1
 
-    def update_state_after_step(self):
+    async def update_state_after_step(self):
         self.state.updated_info = []
         # update metrics especially for cost
         self.state.metrics = self.agent.llm.metrics
+        if self.max_budget_per_task is not None:
+            current_cost = self.state.metrics.accumulated_cost
+            if current_cost > self.max_budget_per_task:
+                await self.report_error(
+                    f'Task budget exceeded. Current cost: {current_cost}, Max budget: {self.max_budget_per_task}'
+                )
+                await self.set_agent_state_to(AgentState.ERROR)
```

Comment:
Is ERROR the state we need for eval?

Reply:
Hmmm yeah in general maybe a better idea is to give users a chance to choose whether they would like to proceed. IIRC we have a similar mechanism for MAX_TOKENS or MAX_TURNS?

But yeah at the moment I believe ERROR is needed for evaluation which is fully automated.

We should probably create an issue for this.

Reply:
UPDATE: just checked that we didn't do so for `MAX_CHARS` and `MAX_ITERATIONS` either...

Reply:
We already have an issue for this: https://github.com/OpenDevin/OpenDevin/issues/1493